File: plugins/plugin-form/python/elizaos_plugin_form/defaults.py
"""
Default value application for forms and controls.

Ensures minimal definitions work correctly by filling in sensible defaults.
"""

from __future__ import annotations

import re
from copy import deepcopy
from typing import cast

from .types import (
    FORM_CONTROL_DEFAULTS,
    FORM_DEFINITION_DEFAULTS,
    FormControl,
    FormDefinition,
    FormDefinitionNudge,
    FormDefinitionTTL,
    FormDefinitionUX,
    JsonValue,
)


def prettify(key: str) -> str:
    """Convert snake_case or kebab-case to Title Case.

    >>> prettify("first_name")
    'First Name'
    >>> prettify("email-address")
    'Email Address'
    """
    spaced = re.sub(r"[-_]", " ", key)
    return re.sub(r"\b\w", lambda m: m.group().upper(), spaced)


def apply_control_defaults(control: dict[str, JsonValue] | FormControl) -> dict[str, JsonValue]:
    """Apply defaults to a control dict and return a complete dict."""
    if isinstance(control, FormControl):
        src: dict[str, JsonValue] = {
            "key": control.key,
            "label": control.label,
            "type": control.type,
            "required": control.required,
        }
        if control.confirm_threshold is not None:
            src["confirm_threshold"] = control.confirm_threshold
        if control.pattern is not None:
            src["pattern"] = control.pattern
        if control.min is not None:
            src["min"] = control.min
        if control.max is not None:
            src["max"] = control.max
        if control.description is not None:
            src["description"] = control.description
    else:
        src = dict(control)

    key = str(src.get("key", ""))
    result: dict[str, JsonValue] = {
        "key": key,
        "label": src.get("label") or prettify(key),
        "type": src.get("type") or cast(str, FORM_CONTROL_DEFAULTS["type"]),
        "required": src.get("required") if src.get("required") is not None else FORM_CONTROL_DEFAULTS["required"],
        "confirm_threshold": (
            src.get("confirm_threshold")
            if src.get("confirm_threshold") is not None
            else FORM_CONTROL_DEFAULTS["confirm_threshold"]
        ),
    }

    # Merge remaining keys from source
    for k, v in src.items():
        if k not in result:
            result[k] = v

    return result


def apply_form_defaults(form: dict[str, JsonValue] | FormDefinition) -> dict[str, JsonValue]:
    """Apply defaults to a form dict and return a complete dict."""
    _def = FORM_DEFINITION_DEFAULTS
    _ux_def: dict[str, JsonValue] = _def["ux"]  # type: ignore[assignment]
    _ttl_def: dict[str, JsonValue] = _def["ttl"]  # type: ignore[assignment]
    _nudge_def: dict[str, JsonValue] = _def["nudge"]  # type: ignore[assignment]

    if isinstance(form, FormDefinition):
        src: dict[str, JsonValue] = {"id": form.id, "name": form.name}
        if form.version is not None:
            src["version"] = form.version
        if form.status is not None:
            src["status"] = form.status
        if form.controls is not None:
            src["controls"] = form.controls  # type: ignore[assignment]
        if form.debug is not None:
            src["debug"] = form.debug
        if form.description is not None:
            src["description"] = form.description
    else:
        src = dict(form)

    form_id = str(src.get("id", ""))

    # UX
    ux_src = src.get("ux") or {}
    if isinstance(ux_src, dict):
        ux = {
            "allow_undo": ux_src.get("allow_undo") if ux_src.get("allow_undo") is not None else _ux_def["allow_undo"],
            "allow_skip": ux_src.get("allow_skip") if ux_src.get("allow_skip") is not None else _ux_def["allow_skip"],
            "max_undo_steps": ux_src.get("max_undo_steps") if ux_src.get("max_undo_steps") is not None else _ux_def["max_undo_steps"],
            "show_examples": ux_src.get("show_examples") if ux_src.get("show_examples") is not None else _ux_def["show_examples"],
            "show_explanations": ux_src.get("show_explanations") if ux_src.get("show_explanations") is not None else _ux_def["show_explanations"],
            "allow_autofill": ux_src.get("allow_autofill") if ux_src.get("allow_autofill") is not None else _ux_def["allow_autofill"],
        }
    else:
        ux = dict(_ux_def)

    # TTL
    ttl_src = src.get("ttl") or {}
    if isinstance(ttl_src, dict):
        ttl = {
            "min_days": ttl_src.get("min_days") if ttl_src.get("min_days") is not None else _ttl_def["min_days"],
            "max_days": ttl_src.get("max_days") if ttl_src.get("max_days") is not None else _ttl_def["max_days"],
            "effort_multiplier": ttl_src.get("effort_multiplier") if ttl_src.get("effort_multiplier") is not None else _ttl_def["effort_multiplier"],
        }
    else:
        ttl = dict(_ttl_def)

    # Nudge
    nudge_src = src.get("nudge") or {}
    if isinstance(nudge_src, dict):
        nudge: dict[str, JsonValue] = {
            "enabled": nudge_src.get("enabled") if nudge_src.get("enabled") is not None else _nudge_def["enabled"],
            "after_inactive_hours": nudge_src.get("after_inactive_hours") if nudge_src.get("after_inactive_hours") is not None else _nudge_def["after_inactive_hours"],
            "max_nudges": nudge_src.get("max_nudges") if nudge_src.get("max_nudges") is not None else _nudge_def["max_nudges"],
        }
        if nudge_src.get("message") is not None:
            nudge["message"] = nudge_src["message"]
    else:
        nudge = dict(_nudge_def)

    # Controls
    raw_controls = src.get("controls") or []
    controls: list[dict[str, JsonValue]] = []
    if isinstance(raw_controls, list):
        for c in raw_controls:
            if isinstance(c, (dict, FormControl)):
                controls.append(apply_control_defaults(c))
            # else skip non-dict items

    result: dict[str, JsonValue] = {
        "id": form_id,
        "name": src.get("name") or prettify(form_id),
        "version": src.get("version") if src.get("version") is not None else _def["version"],
        "status": src.get("status") if src.get("status") is not None else _def["status"],
        "controls": controls,  # type: ignore[dict-item]
        "ux": ux,  # type: ignore[dict-item]
        "ttl": ttl,  # type: ignore[dict-item]
        "nudge": nudge,  # type: ignore[dict-item]
        "debug": src.get("debug") if src.get("debug") is not None else _def["debug"],
    }

    # Merge remaining keys
    for k, v in src.items():
        if k not in result:
            result[k] = v

    return result

File: plugins/plugin-form/rust/src/defaults.rs
//! Default value application for forms and controls.
//!
//! Minimal definitions like `{ id: "contact", controls: [{ key: "email" }] }`
//! get expanded with sensible defaults for all optional fields.

use crate::types::{
    FormControl, FormDefinition, FormDefinitionNudge, FormDefinitionTTL, FormDefinitionUX,
    FormStatus, DEFAULT_CONFIRM_THRESHOLD, DEFAULT_NUDGE_AFTER_INACTIVE_HOURS,
    DEFAULT_NUDGE_MAX_NUDGES, DEFAULT_TTL_EFFORT_MULTIPLIER, DEFAULT_TTL_MAX_DAYS,
    DEFAULT_TTL_MIN_DAYS,
};

/// Convert snake_case or kebab-case to Title Case.
///
/// ```
/// use elizaos_plugin_form::defaults::prettify;
/// assert_eq!(prettify("first_name"), "First Name");
/// assert_eq!(prettify("email-address"), "Email Address");
/// ```
pub fn prettify(key: &str) -> String {
    key.replace(['-', '_'], " ")
        .split_whitespace()
        .map(|word| {
            let mut chars = word.chars();
            match chars.next() {
                Some(c) => {
                    let mut s = c.to_uppercase().to_string();
                    s.extend(chars);
                    s
                }
                None => String::new(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

/// Apply defaults to a FormControl.
pub fn apply_control_defaults(mut control: FormControl) -> FormControl {
    if control.label.is_empty() {
        control.label = prettify(&control.key);
    }
    if control.type_name.is_empty() {
        control.type_name = "text".to_string();
    }
    if control.confirm_threshold.is_none() {
        control.confirm_threshold = Some(DEFAULT_CONFIRM_THRESHOLD);
    }
    control
}

/// Apply defaults to a FormDefinition.
pub fn apply_form_defaults(mut form: FormDefinition) -> FormDefinition {
    if form.name.is_empty() {
        form.name = prettify(&form.id);
    }
    if form.version.is_none() {
        form.version = Some(1);
    }
    if form.status.is_none() {
        form.status = Some(FormStatus::Active);
    }

    // Apply defaults to controls
    form.controls = form.controls.into_iter().map(apply_control_defaults).collect();

    // UX defaults
    let ux = form.ux.get_or_insert(FormDefinitionUX {
        allow_undo: None,
        allow_skip: None,
        max_undo_steps: None,
        show_examples: None,
        show_explanations: None,
        allow_autofill: None,
    });
    if ux.allow_undo.is_none() {
        ux.allow_undo = Some(true);
    }
    if ux.allow_skip.is_none() {
        ux.allow_skip = Some(true);
    }
    if ux.max_undo_steps.is_none() {
        ux.max_undo_steps = Some(5);
    }
    if ux.show_examples.is_none() {
        ux.show_examples = Some(true);
    }
    if ux.show_explanations.is_none() {
        ux.show_explanations = Some(true);
    }
    if ux.allow_autofill.is_none() {
        ux.allow_autofill = Some(true);
    }

    // TTL defaults
    let ttl = form.ttl.get_or_insert(FormDefinitionTTL {
        min_days: None,
        max_days: None,
        effort_multiplier: None,
    });
    if ttl.min_days.is_none() {
        ttl.min_days = Some(DEFAULT_TTL_MIN_DAYS);
    }
    if ttl.max_days.is_none() {
        ttl.max_days = Some(DEFAULT_TTL_MAX_DAYS);
    }
    if ttl.effort_multiplier.is_none() {
        ttl.effort_multiplier = Some(DEFAULT_TTL_EFFORT_MULTIPLIER);
    }

    // Nudge defaults
    let nudge = form.nudge.get_or_insert(FormDefinitionNudge {
        enabled: None,
        after_inactive_hours: None,
        max_nudges: None,
        message: None,
    });
    if nudge.enabled.is_none() {
        nudge.enabled = Some(true);
    }
    if nudge.after_inactive_hours.is_none() {
        nudge.after_inactive_hours = Some(DEFAULT_NUDGE_AFTER_INACTIVE_HOURS);
    }
    if nudge.max_nudges.is_none() {
        nudge.max_nudges = Some(DEFAULT_NUDGE_MAX_NUDGES);
    }

    // Debug default
    if form.debug.is_none() {
        form.debug = Some(false);
    }

    form
}

// ============================================================================
// TESTS
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    // ═══ PRETTIFY ═══

    #[test]
    fn test_prettify_snake_case() {
        assert_eq!(prettify("first_name"), "First Name");
    }

    #[test]
    fn test_prettify_kebab_case() {
        assert_eq!(prettify("email-address"), "Email Address");
    }

    #[test]
    fn test_prettify_single_word() {
        assert_eq!(prettify("email"), "Email");
    }

    #[test]
    fn test_prettify_empty() {
        assert_eq!(prettify(""), "");
    }

    #[test]
    fn test_prettify_multiple_separators() {
        assert_eq!(prettify("full_name-display"), "Full Name Display");
    }

    #[test]
    fn test_prettify_already_capitalized() {
        assert_eq!(prettify("Name"), "Name");
    }

    // ═══ CONTROL DEFAULTS ═══

    #[test]
    fn test_control_defaults_label_from_key() {
        let c = apply_control_defaults(FormControl {
            key: "first_name".to_string(),
            ..Default::default()
        });
        assert_eq!(c.label, "First Name");
    }

    #[test]
    fn test_control_defaults_type_text() {
        let c = apply_control_defaults(FormControl {
            key: "field".to_string(),
            ..Default::default()
        });
        assert_eq!(c.type_name, "text");
    }

    #[test]
    fn test_control_defaults_confirm_threshold() {
        let c = apply_control_defaults(FormControl {
            key: "field".to_string(),
            ..Default::default()
        });
        assert_eq!(c.confirm_threshold, Some(0.8));
    }

    #[test]
    fn test_control_defaults_preserves_label() {
        let c = apply_control_defaults(FormControl {
            key: "field".to_string(),
            label: "Custom Label".to_string(),
            ..Default::default()
        });
        assert_eq!(c.label, "Custom Label");
    }

    #[test]
    fn test_control_defaults_preserves_type() {
        let c = apply_control_defaults(FormControl {
            key: "field".to_string(),
            type_name: "email".to_string(),
            ..Default::default()
        });
        assert_eq!(c.type_name, "email");
    }

    #[test]
    fn test_control_defaults_preserves_threshold() {
        let c = apply_control_defaults(FormControl {
            key: "field".to_string(),
            confirm_threshold: Some(0.95),
            ..Default::default()
        });
        assert_eq!(c.confirm_threshold, Some(0.95));
    }

    // ═══ FORM DEFAULTS ═══

    #[test]
    fn test_form_defaults_name_from_id() {
        let f = apply_form_defaults(FormDefinition {
            id: "user_registration".to_string(),
            controls: vec![],
            ..Default::default()
        });
        assert_eq!(f.name, "User Registration");
    }

    #[test]
    fn test_form_defaults_version() {
        let f = apply_form_defaults(FormDefinition {
            id: "form".to_string(),
            controls: vec![],
            version: None,
            ..Default::default()
        });
        assert_eq!(f.version, Some(1));
    }

    #[test]
    fn test_form_defaults_status() {
        let f = apply_form_defaults(FormDefinition {
            id: "form".to_string(),
            controls: vec![],
            status: None,
            ..Default::default()
        });
        assert_eq!(f.status, Some(FormStatus::Active));
    }

    #[test]
    fn test_form_defaults_ux() {
        let f = apply_form_defaults(FormDefinition {
            id: "form".to_string(),
            controls: vec![],
            ux: None,
            ..Default::default()
        });
        let ux = f.ux.unwrap();
        assert_eq!(ux.allow_undo, Some(true));
        assert_eq!(ux.allow_skip, Some(true));
        assert_eq!(ux.max_undo_steps, Some(5));
        assert_eq!(ux.show_examples, Some(true));
        assert_eq!(ux.show_explanations, Some(true));
        assert_eq!(ux.allow_autofill, Some(true));
    }

    #[test]
    fn test_form_defaults_ttl() {
        let f = apply_form_defaults(FormDefinition {
            id: "form".to_string(),
            controls: vec![],
            ttl: None,
            ..Default::default()
        });
        let ttl = f.ttl.unwrap();
        assert_eq!(ttl.min_days, Some(14.0));
        assert_eq!(ttl.max_days, Some(90.0));
        assert_eq!(ttl.effort_multiplier, Some(0.5));
    }

    #[test]
    fn test_form_defaults_nudge() {
        let f = apply_form_defaults(FormDefinition {
            id: "form".to_string(),
            controls: vec![],
            nudge: None,
            ..Default::default()
        });
        let nudge = f.nudge.unwrap();
        assert_eq!(nudge.enabled, Some(true));
        assert_eq!(nudge.after_inactive_hours, Some(48.0));
        assert_eq!(nudge.max_nudges, Some(3));
    }

    #[test]
    fn test_form_defaults_debug() {
        let f = apply_form_defaults(FormDefinition {
            id: "form".to_string(),
            controls: vec![],
            debug: None,
            ..Default::default()
        });
        assert_eq!(f.debug, Some(false));
    }

    #[test]
    fn test_form_defaults_preserves_name() {
        let f = apply_form_defaults(FormDefinition {
            id: "form".to_string(),
            name: "My Form".to_string(),
            controls: vec![],
            ..Default::default()
        });
        assert_eq!(f.name, "My Form");
    }

    #[test]
    fn test_form_defaults_applies_to_controls() {
        let f = apply_form_defaults(FormDefinition {
            id: "form".to_string(),
            controls: vec![FormControl {
                key: "user_email".to_string(),
                ..Default::default()
            }],
            ..Default::default()
        });
        assert_eq!(f.controls[0].label, "User Email");
        assert_eq!(f.controls[0].confirm_threshold, Some(0.8));
    }

    #[test]
    fn test_form_defaults_partial_ux() {
        let f = apply_form_defaults(FormDefinition {
            id: "form".to_string(),
            controls: vec![],
            ux: Some(FormDefinitionUX {
                allow_undo: Some(false),
                allow_skip: None,
                max_undo_steps: None,
                show_examples: None,
                show_explanations: None,
                allow_autofill: None,
            }),
            ..Default::default()
        });
        let ux = f.ux.unwrap();
        assert_eq!(ux.allow_undo, Some(false)); // preserved
        assert_eq!(ux.allow_skip, Some(true)); // defaulted
    }
}
